modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DataStreamRouterProcessor.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.ingest.common;
+
+import org.elasticsearch.ingest.AbstractProcessor;
+import org.elasticsearch.ingest.ConfigurationUtils;
+import org.elasticsearch.ingest.IngestDocument;
+import org.elasticsearch.ingest.Processor;
+
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.ingest.ConfigurationUtils.newConfigurationException;
+
+public final class DataStreamRouterProcessor extends AbstractProcessor {
+    public static final String TYPE = "data_stream_router";
+
+    private static final String DATA_STREAM_PREFIX = "data_stream.";
+    private static final String DATA_STREAM_TYPE = DATA_STREAM_PREFIX + "type";
+    private static final String DATA_STREAM_DATASET = DATA_STREAM_PREFIX + "dataset";
+    private static final String DATA_STREAM_NAMESPACE = DATA_STREAM_PREFIX + "namespace";
+    private static final String DATA_STREAM_DEFAULT_DATASET = "generic";
+    private static final String DATA_STREAM_DEFAULT_NAMESPACE = "default";
+    private static final String EVENT_DATASET = "event.dataset";
+
+    private static final char[] DISALLOWED_IN_DATASET = new char[] { '\\', '/', '*', '?', '\"', '<', '>', '|', ' ', ',', '#', ':', '-' };
```

Comment:
Can you make sure this is aligned with what we have in the https://github.com/elastic/package-spec for validation? @mtojek Will likely know where to point you to. Same for namespace.

Reply:
@mtojek @jsoriano Can you chime in here? I would have expected that we do some validation.

Reply:
We have this issue open since long ago https://github.com/elastic/package-spec/issues/57, I will ensure this ends up in our backlog.